File: c/0002_Add_Two_Numbers.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

 struct ListNode* createNode(int value) {
    struct ListNode* newNode = (struct ListNode*) malloc(sizeof(struct ListNode));
    newNode->val = value;
    newNode->next = NULL;
    return newNode;
}

struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2) {
    struct ListNode* headNode = createNode(0);
    struct ListNode* linkedList = headNode;
    int carry = 0;

    while (l1 != NULL || l2 != NULL || carry != 0) {
        int digit1 = (l1 != NULL) ? l1->val : 0;
        int digit2 = (l2 != NULL) ? l2->val : 0;

        int sum = digit1 + digit2 + carry;
        int digit = sum % 10;
        carry = sum / 10;

        struct ListNode* node = createNode(digit);
        linkedList->next = node;
        linkedList = linkedList->next;

        l1 = (l1 != NULL) ? l1->next : NULL;
        l2 = (l2 != NULL) ? l2->next : NULL;
    }
    return headNode->next;
}

File: python/0002_Add_Two_Numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        headNode = ListNode(0)
        linkedList = headNode
        carry = 0
        
        while l1 is not None or l2 is not None or carry != 0:
            digit1 = l1.val if l1 is not None else 0
            digit2 = l2.val if l2 is not None else 0
            sum = digit1 + digit2 + carry
            digit = sum % 10
            carry = sum // 10
            node = ListNode(digit)
            linkedList.next = node
            linkedList = linkedList.next
            
            l1 = l1.next if l1 is not None else None
            l2 = l2.next if l2 is not None else None
        
        return headNode.next
